messaging/src/main/java/org/axonframework/messaging/QueueMessageStream.java

```diff
@@ -0,0 +1,171 @@
+     */
+    public QueueMessageStream(BlockingQueue<Entry<M>> queue) {
+        this.queue = queue;
+    }
+
+    /**
+     * Add the given {@code message} and accompanying {@code context} available for reading by a consumer. Any callback
+     * that has been registered with be notified of the availability of a new element.
+     * <p>
+     * If the underling buffer has insufficient space to store the offered element, the method returns {@code false}.
+     *
+     * @param message The message to add to the queue
+     * @param context The context to accompany the message
+     * @return {@code true} if the message was successfully buffered. Otherwise {@code false}
+     */
+    public boolean offer(@Nonnull M message, @Nonnull Context context) {
+        if (queue.offer(new SimpleEntry<>(message, context))) {
+            onAvailableCallbackRef.get().run();
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Marks the queue as completed, indicating to any consumer that no more messages will become available.
+     * <p>
+     * Note that there is no validation on offering items whether the stream is completed. It is the caller's
+     * responsibility to ensure no Messages are {@link #offer(Message, Context) offered} after completion.
+     */
+    public void complete() {
+        completed.set(true);
+        onAvailableCallbackRef.get().run();
+    }
+
+    /**
+     * Marks the queue as completed exceptionally caused by given {@code error}, indicating to any consumer that no more
+     * messages will become available.
+     * <p>
+     * Note that there is no validation on offering items whether the stream is completed. It is the caller's
+     * responsibility to ensure no Messages are {@link #offer(Message, Context) offered} after completion.
+     *
+     * @param error The cause of the exceptional completion
+     */
+    public void complete(Throwable error) {
+        errorRef.set(error);
+        completed.set(true);
+        onAvailableCallbackRef.get().run();
+    }
+
+    /**
+     * Registers given {@code callback} to be invoked when messages have been consumed from the underlying queue. Any
+     * previously registered callback will be replaced.
+     * <p>
+     * The given {@code callback} is also notified when the consumer has requested to {@link #close()} this stream.
+     *
+     * @param callback The callback to invoke when messages are consumed
+     */
+    public void onConsumeCallback(@Nonnull Runnable callback) {
+        this.onConsumeCallback.set(callback);
+    }
+
+    /**
+     * Whether the consumer has requested to {@link #close()} this stream. This is a signal to producing components to
+     * stop emitting more messages, complete the stream, and release any sources associated with this stream.
+     *
+     * @return {@code true} if a close was requested, otherwise {@code false}
```

Comment:
```suggestion
     * Whether the consumer has requested to {@link #close()} this stream. This is a signal to producing components to
     * stop emitting more {@link Message messages}, complete the stream, and release any sources associated with this stream.
     *
     * @return {@code true} if a close was requested, otherwise {@code false}.
```